components/ai_chat/core/browser/engine/engine_consumer_oai.cc

```diff
@@ -0,0 +1,259 @@
+#include "brave/components/ai_chat/core/browser/engine/remote_completion_client.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom-forward.h"
+#include "brave/components/ai_chat/core/common/mojom/ai_chat.mojom.h"
+#include "components/grit/brave_components_strings.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "ui/base/l10n/l10n_util.h"
+
+namespace {
+constexpr char kQuestionPrompt[] =
+    "Propose up to 3 very short questions that a reader may ask about the "
+    "content. Wrap each in <question> tags.";
+}  // namespace
+
+namespace ai_chat {
+
+namespace {
+
+using mojom::CharacterType;
+using mojom::ConversationTurn;
+
+base::Value::List BuildMessages(
+    const std::string& human_input,
+    const std::string& page_content,
+    const std::optional<std::string>& selected_text,
+    const bool& is_video,
+    const EngineConsumer::ConversationHistory& conversation_history) {
+  const std::string prompt_segment_article =
+      page_content.empty()
+          ? ""
+          : base::StrCat(
+                {base::ReplaceStringPlaceholders(
+                     l10n_util::GetStringUTF8(
+                         is_video ? IDS_AI_CHAT_LLAMA2_VIDEO_PROMPT_SEGMENT
+                                  : IDS_AI_CHAT_LLAMA2_ARTICLE_PROMPT_SEGMENT),
+                     {page_content}, nullptr),
+                 "\n\n"});
+
+  std::string date_and_time_string =
+      base::UTF16ToUTF8(TimeFormatFriendlyDateAndTime(base::Time::Now()));
+  std::string system_message = base::StrCat(
+      {prompt_segment_article,
+       base::ReplaceStringPlaceholders(
+           l10n_util::GetStringUTF8(IDS_AI_CHAT_LLAMA2_SYSTEM_MESSAGE_GENERIC),
+           {date_and_time_string}, nullptr)});
+
+  // build
+  base::Value::List messages;
+
+  // append system message
+  {
+    base::Value::Dict message;
+    message.Set("role", "system");
+    message.Set("content", system_message);
+    messages.Append(std::move(message));
+  }
+
+  for (const mojom::ConversationTurnPtr& turn : conversation_history) {
+    base::Value::Dict message;
+    message.Set("role", turn->character_type == CharacterType::HUMAN
+                            ? "user"
+                            : "assistant");
+    message.Set(
+        "content",
+        turn->selected_text
+            ? base::StrCat(
+                  {base::ReplaceStringPlaceholders(
+                       l10n_util::GetStringUTF8(
+                           IDS_AI_CHAT_LLAMA2_SELECTED_TEXT_PROMPT_SEGMENT),
+                       {*turn->selected_text}, nullptr),
+                   "\n\n", turn->text})
+            : turn->text);
+    messages.Append(std::move(message));
+  }
+
+  return messages;
+}
+
+}  // namespace
+
+EngineConsumerOAIRemote::EngineConsumerOAIRemote(
+    const mojom::CustomModelOptions& model_options,
+    scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory) {
+  model_options_ = model_options;
```

Comment:
Just a thought here - since we're implicitly copying the `mojom::CustomModelOptions` here (I think) then isn't it better practice to deal with `mojom::CustomModelOptionsPtr` and clone? Or store the pointer if we know it will definitely be alive longer than the engine consumer (which I don't think we do know since the service which owns it can delete it before this class gets an updated version?).